### PR TITLE
CPBR-1863 add functionality for set_properties and its variants to ub script

### PR DESCRIPTION
**Pull request by hk10111**
### Change Description

add functionality for 
1. setProperties 
2. setPropertiesWithSkipPropCheck
3. setPropertiesWithEnvToPropsWithTwoPrefixes

C3 template changes are part of https://github.com/confluentinc/control-center-next-gen-images/pull/10

### Testing

Added tests in ub_test.go to test various combinations with excludes and required properties

### CODE REVIEW

File: base-java/ub/ub.go [hunk 1]
```diff
@@ -92,6 +92,32 @@ var (
 	}
 )
 
+// Helper function to create slices in templates
+func slice(values ...interface{}) []interface{} {
```
Comment (rahejaprince) [inline base-java/ub/ub.go:96]: where is this needed? is stringSlice method not enough?

File: base-java/ub/ub.go [hunk 2]
```diff
@@ -246,12 +303,57 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
 		if strings.HasPrefix(envName, envPrefix) {
 			trimmedEnvName := strings.TrimPrefix(envName, envPrefix)
 			propertyName := propertyNamePrefix + ConvertKey(trimmedEnvName)
-			config[propertyName] = envValue
+			
+			// Check if the property name is in excludes
```
Comment (rahejaprince) [inline base-java/ub/ub.go:307]: please remove these one line comments which are clear from the code itself.

File: base-java/ub/ub.go [hunk 3]
```diff
@@ -246,12 +303,57 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
...
+			if slices.Contains(excludeEnvs, propertyName) {
+				continue
+			}
+			
+			// Check if the property name starts with any of the excluded prefixes
```
Comment (rahejaprince) [inline base-java/ub/ub.go:312]: please remove this comment also.

File: base-java/ub/ub.go [hunk 4]
```diff
@@ -246,12 +303,57 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
...
+func setProperties(properties map[string][]string, required bool, excludes []string) map[string]string {
+	result := make(map[string]string)
+	env := GetEnvironment()
+
+	for property, ks := range properties {
+		var nsResult string
+		// Find first non-empty value
+		for _, k := range ks {
+			if slices.Contains(excludes, k) {
+				// If the key is in excludes, set empty string and break
```
Comment (rahejaprince) [inline base-java/ub/ub.go:337]: please remove this comment

File: base-java/ub/ub.go [hunk 5]
```diff
@@ -246,12 +303,57 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
...
 		}
 	}
 	return config
 }
 
+func setProperties(properties map[string][]string, required bool, excludes []string) map[string]string {
+	result := make(map[string]string)
+	env := GetEnvironment()
+
+	for property, ks := range properties {
```
Comment (rahejaprince) [inline base-java/ub/ub.go:332]: let's use better variable names `ks` doesn't convey what this is.

File: base-java/ub/ub.go [hunk 6]
```diff
@@ -246,12 +303,57 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
...
+		var nsResult string
+		// Find first non-empty value
+		for _, k := range ks {
```
Comment (rahejaprince) [inline base-java/ub/ub.go:335]: please use a better variable name for `k`

File: base-java/ub/ub.go [hunk 7]
```diff
@@ -246,12 +303,57 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
...
+				nsResult = ""
+				break
+			}
+			if val, exists := env[k]; exists {
+				nsResult = val
+				break
+			}
+		}
+
+		// Fill the map based on conditions
```
Comment (rahejaprince) [inline base-java/ub/ub.go:347]: please remove these comments.

File: base-java/ub/ub_test.go [hunk 8]
```diff
@@ -537,11 +537,148 @@ func TestEnvToProps(t *testing.T) {
 					os.Unsetenv(k)
 				}
 			}()
-			result := envToProps(tt.envPrefix, tt.propPrefix, tt.exclude)
+			result := envToProps(tt.envPrefix, tt.propPrefix, tt.exclude, nil)
```
Comment (rahejaprince) [inline base-java/ub/ub_test.go:540]: let's add a test case where the value is not nil. Also add this new parameter added to the test as part of test input instead of hardcoding to nil here.

File: base-java/ub/ub_test.go [hunk 9]
```diff
@@ -537,11 +537,148 @@ func TestEnvToProps(t *testing.T) {
...
+func Test_setPropertiesWithEnvToPropsWithTwoPrefixes(t *testing.T) {
+	type args struct {
+		primaryEnvPrefix   string
+		secondaryEnvPrefix string
+		propPrefix         string
+		excludes           []string
+	}
+
+	// Set up test environment variables
+	os.Setenv("PRIMARY_TEST1", "value1")
```
Comment (rahejaprince) [inline base-java/ub/ub_test.go:558]: we should reset these properties in the end to make sure other tests are not broken because of this.

File: base-java/ub/ub.go [hunk 10]
```diff
@@ -246,12 +297,52 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
...
 	return config
 }
 
+func setProperties(properties map[string][]string, required bool, excludes []string) map[string]string {
+	result := make(map[string]string)
+	env := GetEnvironment()
+
+	for property, propertyTranslationList := range properties {
+		var nsResult string
+		// Find first non-empty value
```
Comment (rahejaprince) [inline base-java/ub/ub.go:325]: please remove this comment.

File: base-java/ub/ub.go [hunk 11]
```diff
@@ -246,12 +297,52 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
...
+			if slices.Contains(excludes, envVar) {
+				nsResult = ""
+				break
+			}
+			if val, exists := env[envVar]; exists {
+				nsResult = val
+				break
+			}
+		}
+
```
Comment (rahejaprince) [inline base-java/ub/ub.go:336]: nit: please remove redundant newlines

File: base-java/ub/ub_test.go [hunk 12]
```diff
@@ -537,11 +558,159 @@ func TestEnvToProps(t *testing.T) {
...
+}
+
+func Test_setProperties(t *testing.T) {
+	type args struct {
+		properties map[string][]string
+		required   bool
+		excludes   []string
+	}
+
+	// Set up test environment variables
```
Comment (rahejaprince) [inline base-java/ub/ub_test.go:645]: nit: please remove these comments.

File: base-java/ub/ub_test.go [hunk 13]
```diff
@@ -537,11 +558,159 @@ func TestEnvToProps(t *testing.T) {
...
+	os.Setenv("TEST1", "value1")
```
Comment (rahejaprince) [inline base-java/ub/ub_test.go:646]: let's use a consistent pattern to set environment variables.
Also write more extensive tests where there are multiple combination of environment variables and corresponding outputs are defined.

File: base-java/ub/ub_test.go [hunk 14]
```diff
@@ -537,11 +558,159 @@ func TestEnvToProps(t *testing.T) {
...
+func Test_setPropertiesWithEnvToPropsWithTwoPrefixes(t *testing.T) {
+	type args struct {
+		primaryEnvPrefix   string
+		secondaryEnvPrefix string
+		propPrefix         string
+		excludes           []string
+	}
+
+	// Set up test environment variables
+	envVars := map[string]string{
```
Comment (rahejaprince) [inline base-java/ub/ub_test.go:579]: let's write more extensive tests where there are multiple combination of environment variables and corresponding outputs are defined.

File: base-java/ub/ub.go [hunk 15]
```diff
@@ -246,12 +288,50 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
 		if strings.HasPrefix(envName, envPrefix) {
 			trimmedEnvName := strings.TrimPrefix(envName, envPrefix)
 			propertyName := propertyNamePrefix + ConvertKey(trimmedEnvName)
-			config[propertyName] = envValue
+
+			if slices.Contains(excludeEnvs, propertyName) {
```
Comment (rahejaprince) [inline base-java/ub/ub.go:292]: nit: will `excludeEnvs` has strings in the format of `propertyName`? propertyName is actual property name (e.g. a.b.c=x and env are of format A_B_C

This seems to be a redundant check, please verify

File: base-java/ub/ub.go [hunk 16]
```diff
@@ -246,12 +288,50 @@ func envToProps(envPrefix, propertyNamePrefix string, excludeEnvs []string) map[
...
+			}
+			if !shouldExclude {
+				config[propertyName] = envValue
+			}
 		}
 	}
 	return config
 }
 
+func setProperties(properties map[string][]string, required bool, excludes []string) map[string]string {
```
Comment (rahejaprince) [inline base-java/ub/ub.go:310]: nit: let's rename excludes to `excludeEnvs`, this will be more descriptive. From the code it seems this is a list of environment variables that have to be excluded.

File: base-java/ub/ub.go (hunk 15 above)
Comment (hk10111) [reply base-java/ub/ub.go:292]: @rahejaprince this check is preventing duplicate entries in the properties file

File: base-java/ub/ub.go [hunk 17]
```diff
@@ -263,10 +263,10 @@ func buildProperties(spec ConfigSpec, environment map[string]string) map[string]
 	return config
 }
 
-func setPropertiesWithEnvToPropsWithTwoPrefixes(primaryEnvPrefix, secondaryEnvPrefix, propPrefix string, excludes []string) map[string]string {
+func setPropertiesWithEnvToPropsWithTwoPrefixes(primaryEnvPrefix, secondaryEnvPrefix, propPrefix string, excludes []string, excludeProps []string) map[string]string {
```
Comment (rahejaprince) [inline base-java/ub/ub.go:266]: lets rename `excludes` to `excludeEnvs` for consistency and more clarity

Comment (hk10111) [reply base-java/ub/ub.go]: fixed

**Review by rahejaprince — DISMISSED** [review]

**Review by rahejaprince — APPROVED** [review]
